lib/src/react_client/chain_refs.dart

```diff
@@ -0,0 +1,87 @@
+///
+/// Inputs can be callback refs, [createRef]-based refs ([Ref]),
+/// or raw JS `createRef` refs ([JsRef]).
+///
+/// If either [ref1] or [ref2] is null, the other ref will be passed through.
+///
+/// If both are null, null is returned.
+///
+/// ```dart
+/// final Example = forwardRef((props, ref) {
+///   final localRef = useRef<FooComponent>();
+///   return Foo({
+///     ...props,
+///     'ref': chainRefs(localRef, ref),
+///   });
+/// });
+/// ```
+dynamic chainRefs(dynamic ref1, dynamic ref2) {
+  return chainRefList([ref1, ref2]);
+}
+
+/// Like [chainRefs], but takes in a list of [refs].
+dynamic chainRefList(List<dynamic> refs) {
+  // Wrap in an assert so iteration doesn't take place unnecessarily
+  assert(() {
+    refs.forEach(_validateChainRefsArg);
+    return true;
+  }());
+
+  final nonNullRefs = refs.where((ref) => ref != null).toList();
+
+  // Return null if there are no refs to chain
+  if (nonNullRefs.isEmpty) return null;
+
+  // Pass through the ref if there's nothing to chain it with
+  if (nonNullRefs.length == 1) return nonNullRefs[0];
+
+  // Adapted from https://github.com/smooth-code/react-merge-refs/tree/v1.0.0
+  //
+  // Copyright 2019 Smooth Code
+  //
+  // Permission is hereby granted, free of charge, to any person obtaining a
+  // copy of this software and associated documentation files (the 'Software'),
+  // to deal in the Software without restriction, including without limitation
+  // the rights to use, copy, modify, merge, publish, distribute, sublicense,
+  // and/or sell copies of the Software, and to permit persons to whom the
+  // Software is furnished to do so, subject to the following conditions:
+  //
+  // The above copyright notice and this permission notice shall be included in
+  // all copies or substantial portions of the Software.
+  //
+  // THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+  // IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+  // FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+  // THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+  // LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+  // FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+  // DEALINGS IN THE SOFTWARE.
+  void _chainedRef(value) {
+    for (final ref in nonNullRefs) {
+      if (ref is Function) {
+        ref(value);
+      } else if (ref is Ref) {
+        // ignore: invalid_use_of_protected_member
+        ref.current = value;
+      } else {
+        // ignore: invalid_use_of_protected_member
+        (ref as JsRef).current = value;
+      }
+    }
+  }
+
+  return _chainedRef;
+}
+
+void _validateChainRefsArg(dynamic ref) {
+  assert(ref is! Function || ref is Function(Null), 'callback refs must take a single argument');
+  assert(ref is! String, 'String refs cannot be chained');
+}
```

Comment:
Can we assert here that `ref` is either a `Ref` or a callback ref? Or could `ref` be other types? It just seems weird that if I pass in something like an integer in the list of refs there are no failures.

Reply:
Unfortunately we can't because type-checking on `JsRef` will return `true` for all cases, at least last time I checked. 

I'll try adding more checks to see if that works now, and if not, I'll add a comment with more info

Reply:
Addressed in ce82545 and also added better test coverage for JS refs, which I discovered weren't really being tested fully